portfolio/src/main/java/com/google/sps/servlets/FormServlet.java

```diff
@@ -106,4 +126,39 @@ private String convertToJson(ArrayList<Comment> list) {
     String json = gson.toJson(list);
     return json;
   }
+
+  // Returns a URL that points to the uploaded file, or null if the user didn't upload a file
+  private String getUploadedFileUrl(HttpServletRequest request, String formInputElementName) {
+    BlobstoreService blobstoreService = BlobstoreServiceFactory.getBlobstoreService();
+    Map<String, List<BlobKey>> blobs = blobstoreService.getUploads(request);
+    List<BlobKey> blobKeys = blobs.get("image");
+
+    // Return null if user submits form without selecting a file (dev server)
+    if (blobKeys == null || blobKeys.isEmpty()) {
+      return null;
+    }
+
+    // Get the key of the file input
+    BlobKey blobKey = blobKeys.get(0);
+
+    // Return null if user submits form without selecting a file (live server)
+    BlobInfo blobInfo = new BlobInfoFactory().loadBlobInfo(blobKey);
+    if (blobInfo.getSize() == 0) {
+      blobstoreService.delete(blobKey);
+      return null;
+    }
+
+    // Use ImagesService to get a URL that points to the uploaded file
+    ImagesService imagesService = ImagesServiceFactory.getImagesService();
+    ServingUrlOptions options = ServingUrlOptions.Builder.withBlobKey(blobKey);
+
+    // To support running in Google Cloud Shell with AppEngine's devserver, we must use the relative
+    // path to the image, rather than the path returned by imagesService which contains a host
```

Comment:
Is this what the MalformedURLException is catching? On the devserver it gives an invalid URL?

Reply:
Yes I believe so--the devserver uses the 'try' portion of the 'try-catch' as it needs a path, whereas the live server would use the 'catch' portion as it would need an actual URL (I have not been able to confirm the latter half true as there is no Blobstore in production, but this is what I came to understand after going through the tutorial and looking over what the functions do).